File: src/main/java/destiny/astrology/classical/rules/IRule.java
/**
 * @author smallufo
 * Created on 2007/12/29 at 上午 4:09:26
 */
package destiny.astrology.classical.rules;

import destiny.astrology.IHoroscopeModel;
import destiny.astrology.Planet;
import kotlin.Pair;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Locale;
import java.util.function.Predicate;

public interface IRule extends Predicate<Pair<Planet, IHoroscopeModel>> {

  @Override
  default boolean test(Pair<Planet, IHoroscopeModel> t) {
    return isApplicable(t.getFirst() , t.getSecond());
  }

  boolean isApplicable(Planet planet, IHoroscopeModel h);

  String getName();

  String getName(Locale locale);

  /** 取得某 Locale 之下的註解 */
  @Nullable
  String getComment(Planet planet , IHoroscopeModel h , @NotNull Locale locale);

  @Nullable
  default String getComment(Planet planet , IHoroscopeModel h) {
    return getComment(planet , h , Locale.getDefault());
  }
}

File: src/main/java/destiny/astrology/classical/rules/IRule.kt
/**
 * @author smallufo
 * Created on 2007/12/29 at 上午 4:09:26
 */
package destiny.astrology.classical.rules

import destiny.astrology.IHoroscopeModel
import destiny.astrology.Planet
import java.util.*
import java.util.function.Predicate

enum class RuleType {
  ESSENTIAL,
  ACCIDENTAL,
  DEBILITY
}


interface IRule : Predicate<Pair<Planet, IHoroscopeModel>> {

  val name: String

  val type : RuleType

  override fun test(t: Pair<Planet, IHoroscopeModel>): Boolean {
    return isApplicable(t.first, t.second)
  }

  fun isApplicable(planet: Planet, h: IHoroscopeModel): Boolean

  fun getName(locale: Locale): String

  /** 取得某 Locale 之下的註解  */
  fun getComment(planet: Planet, h: IHoroscopeModel, locale: Locale): String?

  fun getComment(planet: Planet, h: IHoroscopeModel): String? {
    return getComment(planet, h, Locale.getDefault())
  }
}
